segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1350 @@
+                            segmentMetadata.setOwnershipChanged(false);
+                            log.debug("write: First write after failover. segment={}", streamSegmentName);
+                        }
+                        didSegmentLayoutChange = true;
+                        log.debug("write: New chunk added. segment={} chunk={} offset={} ", streamSegmentName, newChunkName, segmentMetadata.getLength());
+                    } else {
+                        // No new chunk needed just write data to existing chunk.
+                        chunkWrittenMetadata = lastChunkMetadata;
+                        chunkHandle = chunkStorage.openWrite(lastChunkMetadata.getName());
+                    }
+
+                    // Calculate the data that needs to be written.
+                    long offsetToWriteAt = currentOffset - segmentMetadata.getLastChunkStartOffset();
+                    int bytesToWrite = (int) Math.min(bytesRemaining, segmentMetadata.getMaxRollinglength() - offsetToWriteAt);
+                    Preconditions.checkState(0 != bytesToWrite, "Attempt to write zero bytes");
+
+                    try {
+                        int bytesWritten;
+                        // Finally write the data.
+                        try (BoundedInputStream bis = new BoundedInputStream(data, bytesToWrite)) {
+                            bytesWritten = chunkStorage.write(chunkHandle, offsetToWriteAt, bytesToWrite, bis);
+                        }
+
+                        // Update the counts
+                        bytesRemaining -= bytesWritten;
+                        currentOffset += bytesWritten;
+
+                        // Update the metadata for segment and chunk.
+                        Preconditions.checkState(bytesWritten >= 0);
+                        segmentMetadata.setLength(segmentMetadata.getLength() + bytesWritten);
+                        chunkWrittenMetadata.setLength(chunkWrittenMetadata.getLength() + bytesWritten);
+                        txn.update(chunkWrittenMetadata);
+                        txn.update(segmentMetadata);
+                    } catch (IndexOutOfBoundsException e) {
+                        throw new BadOffsetException(streamSegmentName, chunkStorage.getInfo(chunkHandle.getChunkName()).getLength(), offset);
+                    }
+                }
+
+                // Check invariants.
+                segmentMetadata.checkInvariants();
+
+                // if layout did not change then commit with lazyWrite.
+                txn.commit(!didSegmentLayoutChange);
+                isCommited = true;
+
+                // Post commit actions.
+                if (isStorageSystemSegment(streamSegmentName) && chunksAddedCount > 0) {
+                    // commit all system log records.
+                    Preconditions.checkState(chunksAddedCount == systemLogRecords.size());
+                    systemJournal.commitRecords(systemLogRecords);
```

Comment:
What happens in case of failure here ?
the chunkStorage already performed the write, there is no transaction that spans the two entities.

Reply:
Good catch ! All segment metadata is now stored in some key-value store (ChunkMetadataStore). 
However that metadata itself needs to store its data in system segments. 
To avoid circular reference (having metadata about system segments stored in system segments), changes to system segments metadata is logged in systemJournal and metadata about system segments is not written to the metadata table itself. 
You are correct that writing to these logs might actually fail even if the transaction is committed. 

I think I need to write systemJournal record before committing. (Technically we don't write the metadata about system segments to underlying store at all, so that commit can't fail because of any Io errors.)

Reply:
fixed.

Reply:
fixed.